provisioner/data.js

```diff
@@ -1,30 +1,704 @@
+      return workerType.killAll(debug);
+    }));
+  });
+
+  return p;
+};
+
+
+/**
+ * Load a single workerType by name
+ */
+WorkerType.load = function(workerType) {
+  assert(workerType);
+  return base.Entity.load.call(this, {
+    workerType: workerType
+  });
+};
+
+
+/**
+ * Return an Object for JSON encoding which represents
+ * the data associated with this WorkerType.  This is a
+ * method intended for use in displaying the data associated
+ * with a given workerType
+ */
+WorkerType.prototype.json = function() {
+  return lodash.clone(this.__properties);
+};
+
+
+/**
+ * Return the list of regions which can be operated in.
+ * This is a subset of the regions which a given node
+ * is allowed to run in and the list of regions that the
+ * API is configred to run in.
+ */
+WorkerType.prototype.listRegions = function() {
+  var that = this;
+  return Object.keys(this.regions).filter(function(region) {
+    return that.ec2.regions.indexOf(region) !== -1;
+  }); 
+};
+
+
+/**
+ * We use KeyPair names to determine ownership and workerType
+ * in the EC2 world because we can't tag SpotRequests until they've
+ * mutated into Instances.  This sucks and all, but hey, what else
+ * can we do?  This method checks which regions have the required
+ * KeyPair already and creates the KeyPair in regions which do not
+ * already have it
+ */
+WorkerType.prototype.createKeyPair = function() {
+  var that = this;
+
+  var keyName = this.keyPrefix + this.workerType;
+
+  var p = this.ec2.describeKeyPairs.inRegions(this.listRegions(), {
+    Filters: [{
+      Name: 'key-name',
+      Values: [keyName]
+    }] 
+  });
+
+  p = p.then(function(res) {
+    var toCreate = [];
+
+    that.listRegions().forEach(function(region) {
+      var matchingKey = res[region].KeyPairs[0];
+      if (!matchingKey) {
+        toCreate.push(that.ec2.importKeyPair.inRegion(region, {
+          KeyName: keyName,
+          PublicKeyMaterial: that.pubKey,
+        }));
+      } 
+    });
+    return Promise.all(toCreate);
+  });
+
+  return p;
+
+};
+
+
+/**
+ * Delete a KeyPair when it's no longer needed.  This method
+ * does nothing more and you shouldn't run it until you've turned
+ * everything off.
+ */
+WorkerType.prototype.deleteKeyPair = function() {
+  var that = this;
+
+  var keyName = this.keyPrefix + this.workerType;
+
+  var p = this.ec2.describeKeyPairs({
+    Filters: [{
+      Name: 'key-name',
+      Values: [keyName]
+    }] 
+  });
+
+  p = p.then(function(res) {
+    var toDelete = [];
+
+    that.listRegions().forEach(function(region) {
+      var matchingKey = res[region].KeyPairs[0];
+      if (matchingKey) {
+        toDelete.push(that.ec2.deleteKeyPair.inRegion(region, {
+          KeyName: keyName,
+        }));
+      } 
+    });
+    return Promise.all(toDelete);
+  });
+
+  return p;
+
+};
+
+/**
+ * Shutdown all instances of this workerType, cancel
+ * any open spot requests.
+ */
+WorkerType.prototype.killAll = function(debug) {
```

Comment:
I'm not sure when you use this... But it seems tempting to me to do it in `AwsState`..
If it's for the case where you delete the workerType, it seems to me like it's not a special case.

I might not have made this clear. But I consider the aws-provisioner to own all ec2 instances that has a keyName starting with `aws-provisioner-managed-` (or whatever prefix the provisioner is configured with). And we should terminate spot requests and instances with this key, but an invalid `workerType` ie. a workerType that doesn't exist because it was deleted etc...